src/NuGet.Core/NuGet.Build.Tasks.Console/Program.cs

```diff
@@ -144,23 +147,62 @@ private static bool IsDebug()
+        /// <param name="getStream">A <see cref="Func{TResult}" /> that is called to get a <see cref="Stream" /> to read.</param>
+        /// <param name="errorWriter">A <see cref="TextWriter" /> to write errors to.</param>
         /// <param name="arguments">A <see cref="T:Tuple&lt;Dictionary&lt;string, string&gt;, FileInfo, string, Dictionary&lt;string, string&gt;&gt;" /> that receives the parsed command-line arguments.</param>
         /// <returns><code>true</code> if the arguments were successfully parsed, otherwise <code>false</code>.</returns>
-        private static bool TryParseArguments(string[] args, out (Dictionary<string, string> Options, FileInfo MSBuildExeFilePath, string EntryProjectFilePath, Dictionary<string, string> MSBuildGlobalProperties) arguments)
+        internal static bool TryParseArguments(string[] args, Func<Stream> getStream, TextWriter errorWriter, out (Dictionary<string, string> Options, FileInfo MSBuildExeFilePath, string EntryProjectFilePath, Dictionary<string, string> MSBuildGlobalProperties) arguments)
         {
-            if (args.Length != 4)
+            arguments = (null, null, null, null);
+
+            // This application receives 3 or 4 arguments:
+            // 1. A semicolon delimited list of key value pairs that are the options to the program
+            // 2. The full path to MSBuild.exe
+            // 3. The full path to the entry project file
+            // 4. (optional) A semicolon delimited list of key value pairs that are the global properties to pass to MSBuild
+            if (args.Length < 3 || args.Length > 4)
             {
-                arguments = (null, null, null, null);
+                // An error occurred parsing command-line arguments in static graph-based restore as there was an unexpected number of arguments, {0}. Please file an issue at https://github.com/NuGet/Home. {0}
+                LogError(errorWriter, Strings.Error_StaticGraphRestoreArgumentParsingFailedInvalidNumberOfArguments, args.Length);
 
                 return false;
             }
 
             try
             {
-                var options = ParseSemicolonDelimitedListOfKeyValuePairs(args[0]);
+                Dictionary<string, string> options = ParseSemicolonDelimitedListOfKeyValuePairs(args[0]);
                 var msbuildExeFilePath = new FileInfo(args[1]);
                 var entryProjectFilePath = args[2];
-                var globalProperties = ParseSemicolonDelimitedListOfKeyValuePairs(args[3]);
+
+                Dictionary<string, string> globalProperties = null;
+
+                // If there are 3 arguments then the global properties will be read from STDIN
+                if (args.Length == 3)
+                {
+#if NETFRAMEWORK
+                    if (AppDomain.CurrentDomain.IsDefaultAppDomain())
+                    {
+                        // The application is called twice and the first invocation does not need to read the global properties
```

Comment:
I am curious to understand why this application would be called twice.

Reply:
When the application starts the first time on .NET Framework, its own `App.config` is loaded.  However, later on its going to load MSBuild assemblies and run targets/tasks which come from the .NET SDK.  When you host MSBuild in your process, you miss out on all of the assemblies with binding redirects.  So this application launches itself again in a new AppDomain with MSBuild's `App.config` instead which makes the loader use the binding redirects and makes it possible to host builds in your process.

Reply:
This answer should be an inline comment.

Reply:
This information is in the code just not visible in this diff: https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.Build.Tasks.Console/Program.cs#L62-L65